asllib/doc/ASLTypingReference.tex

```diff
@@ -5859,15 +5973,1111 @@ \subsection{Code}
+    \staticeval(\tenv, \veone) \typearrow \vvone\\
+    \tododefine{bitvector\_length}(\bv) > \posmax\\
+    \vv \eqdef \lbitvector(\tododefine{extract\_slice}(\bv, \positions))
+  }
+  {
+    \staticeval(\tenv, \ESlice(\lbitvector(\bv), \slices)) \typearrow \vv
+  }
+  \and
+  \inferrule[e\_slice\_type\_error]{
+    \astlabel(\veone) \not\in \{\lint, \lbitvector\}
+  }
+  {
+    \staticeval(\tenv, \ESlice(\veone, \slices)) \typearrow \TypeError(\texttt{TypeMismatch}(\veone, [\TInt, \TBits]))
+  }
+  \and
+  \inferrule[e\_cond]{
+    \staticeval(\tenv, \econd) \typearrow \vcond\\
+    \vcond \eqname \lbool(\vb)\\
+    \vep \eqdef \choice{\vb}{\veone}{\vetwo}\\
+    \staticeval(\tenv, \vep) \typearrow \vv
+  }
+  {
+    \staticeval(\tenv, \ECond(\econd, \veone, \vetwo)) \typearrow \vv
+  }
+\end{mathpar}
+\end{emptyformal}
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+\chapter{Equivalence Testing}
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+In this chapter, we define \emph{conservative} equivalence tests for expressions and types.
+By conservative, we mean that if a test returns $\True$ then the expressions or types being
+compared are indeed equivalent, but if the test returns $\False$ then
+there are two possibilities:
+\begin{itemize}
+  \item the expressions or types are not equivalent;
+  \item the expressions or types are equivalent, but the reasoning power of our rules
+  is not enough to prove it, and so we ``pessimistically'' answer negatively.
+\end{itemize}
+In proof-theoretic terms, we say that such tests are \emph{sound} but \emph{incomplete}.
+
+Notice that for a conservative test, it is always correct to return $\False$.
+
+\begin{definition}[\hypertarget{def-equal}{Equal Rule}]
+  The meta function $\equate$ equates to values of the same mathematical type and returns
+  the respective Boolean value:
+  \begin{mathpar}
+      \inferrule[equate-true]{a = b}
+      {
+          \equate(a, b) \rightarrow \True
+      }
+      \and
+      \inferrule[equate-false]{a \neq b}
+      {
+          \equate(a, b) \rightarrow \False
+      }
+  \end{mathpar}
+  \end{definition}
+  The reason we define such a function for equality is to allow using it as a short-circuiting premise.
+
+\section{TypingRule.ExprEqual}
+The function
+\[
+  \exprequal(\overname{\staticenvs}{\tenv} \aslsep \overname{\expr}{\veone} \aslsep \overname{\expr}{\vetwo}) \rightarrow \overname{\{\True, \False\}}{\vb}
+\]
+conservatively checks whether the expression $\veone$ is equivalent to the expression $\vetwo$ in environment $\tenv$.
+
+\begin{emptyformal}
+\subsection{Formally}
+\begin{mathpar}
+  \inferrule{
+    \exprequalnorm(\tenv, \veone, \vetwo) \typearrow \False \terminateas \True\\\\
+    \exprequalcase(\tenv, \veone, \vetwo) \typearrow \vb
+  }
+  {
+    \exprequal(\tenv, \veone, \vetwo) \typearrow \vb
+  }
+\end{mathpar}
+\end{emptyformal}
```

Comment:
I fixed the link issue.

Reply:
> I'm afraid I really think that you should write the corresponding english, or at least expand the inference rule templace because this is unreadable for a new-comer.
Right. I intend to add a Prose subsection and also explain how I use the macro rules there.

>Also I think one or two sentences to explain what each of those function does would be useful, as well as an >explanation of why they are named norm and case
Will do.

Reply:
deferred to next PR.